### PR TITLE
Initial implementation of check/collate "col_attrs", need some help t…

### PR DESCRIPTION
**Pull request by mingodad**
Initial implementation of check/collate "col_attrs", need some help to output 'check' expressions to json, see the modified src/demo.
The same problem I have now to output check expression, also apply to default that actually only handle str_literal or num_literal but according to https://sqlite.org/lang_createtable.html it can also be an expression.

### CODE REVIEW

File: sources/cg_json_schema.c [hunk 1]
```diff
@@ -383,6 +383,18 @@ static void cg_json_col_attrs(charbuf *output, col_info *info) {
     cg_json_default_value(output, sem_get_col_default_value(info->attrs));
   }
 
+  // check/collate are not setting sem_type flags, we have to walk the list of attributes and gather them all
```
[the remark this answers is not in the document]
Comment (ricomariani) [reply sources/cg_json_schema.c:386]: Note you can make your own buffer for the output as well.  For instance

```
static void cg_json_action(charbuf *output, int32_t action) {
  CHARBUF_OPEN(sql);    <--- makes a local buffer
  gen_set_output_buffer(&sql);
  if (!action) {
    action = FK_NO_ACTION;
  }
  gen_fk_action(action); <-- writes an fk action into it
  bprintf(output, "\"%s\"", sql.ptr);  <-- emits it quoted
  CHARBUF_CLOSE(sql);  <-- discard the buffer
}
```
Pretty much everything is these charbufs

**Comment by ricomariani** [comment]
I sort of punted on expressions inside of the default value because

1. they almost never come up
2. the rules for those expressions are going to be a bit strange, they really have to be constant expressions to make any sense.  Normal expressions allow for (e.g.)  (select x from y) and that is just a horrible idea here.  I didn't want to get into the business of analyzing expressions trees for "constant-ness" and so I punted...

**Comment by mingodad** [comment]
Thanks for reply and the help/feedback !

There is also date/datetime types and their corresponding default values `CURRENT_DATE` and `CURRENT_TIMESTAMP` that are `name` and is not managed right now, also indexes can have expressions.

For the `json` output of expressions I think that outputting the two representations would be nice (text/tree).

The 80/20 rule seems to be acting here you already got the 80% done, now the last 20% will take 80% of the development time.

Just to register other attempts to creating stored procedures for sqlite:
- https://www.sqliteconcepts.org/pl_proc.html
- https://www.gaia-gis.it/fossil/libspatialite/wiki?name=Stored+Procedures
- https://github.com/wolfch/sqlite-3.7.3.p1/blob/sproc-1/ReadMeStoredProcs.txt
- https://github.com/bloomberg/comdb2

Other stored procedure attempts:
- https://github.com/okbob/plpgsql_check

Cheers !

**Comment by ricomariani** [comment]
I had "most" of the SQLite language done about 2 months after I started.  Now it's been 2 years :D  I think it's the 80/2000 rule :D

We aren't emitting expression trees anywhere else yet, only in diagnostic output.  If we were going to do this in JSON I'd do it as a json object kind of thing that's isomorphic to the AST.   The trouble is this isn't stable as the AST changes over time pretty significantly.  There aren't a lot of customers for this sort of thing anyway.  What people do with the JSON is examine the external contracts to the stored procedures so they can make wrappers or other codegen downstream.  The Java and Objc output should have been done this way but the JSON didn't exist at that time...

**Comment by ricomariani** [comment]
If you don't feel comfortable finishing this up, this is worthy and I can take it the rest of way without much effort.  I don't want to steal your thunder though.  Just let me know.

**Comment by mingodad** [comment]
Go ahead, thank you for it, all I did was intended to help.

**Comment by ricomariani** [comment]
OK I'll take this over the finish line.  No further updates to the PR are needed.    :D

**Comment by ricomariani** [comment]
Note: I've been keeping this PR fresh (I have it merged to head still).  I still plan to do this soon but I've been kinda busy this week :D

**Comment by ricomariani** [comment]
This is about to land

**Pull request closed without merging**